### PR TITLE
Add a way to configure navigation to the owner (WithOwner)

### LINKED ISSUE #9148: Metadata: Configuring class as EntityType & later as owned entity type explicitly should throw
```C#
public class Blog
{
    public int Id { get; set; }
    public BlogDetails BlogDetails { get; set; }
}

public class BlogDetails
{
    public int Value { get; set; }
}

protected override void OnModelCreating(ModelBuilder modelBuilder)
{
    // Configure model
    modelBuilder.Entity<BlogDetails>();
    modelBuilder.Entity<Blog>().OwnsOne(e => e.BlogDetails);
}
```

In above model `BlogDetails` has been configured as entity type first and then used as owned entity type. At present it works and creates the own entity type. It should throw exception because it is conflict in Explicit configuration source (which is what our behavior is for most other fluent APIs).
Also if this does not throw (and silently remove entity type and add owned entity type) then any facet configuration done on entity types are lost which may seem confusing to user since user may understand `.EntityType<>` as a replacement of `.ComplexType<>` may be.

found while investigating issue #9144

### LINKED ISSUE #12444: Add API for configuring inverse from an owned type
With this we can disallow using `HasOne`/`HasMany` to configure existing ownerships, except for old snapshots.

Also remove `HasMany` from the owned type configuration builder.

### LINKED ISSUE #14153: Set() should throw for owned types
From https://github.com/aspnet/EntityFramework.Docs/issues/1168 by @voroninp

```CSharp
using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.EntityFrameworkCore;

namespace EfCore22.OwnedTypes
{
    class Program
    {
        static void Main(string[] args)
        {
            using (var ctx = new TestContext())
            {
                ctx.Database.EnsureCreated();
                ctx.Add(new Order { Items = { new OrderItem { Name = "Just an item." } } });
                ctx.SaveChanges();
            }

            using (var ctx = new TestContext())
            {
                var items = ctx.Set<OrderItem>().ToList();
                Console.WriteLine($"I managed to successfully get '{items.Count}' order items.");
            }

            Console.ReadKey();
        }
    }

    public class Order
    {
        public int Id { get; set; }
        public List<OrderItem> Items { get; } = new List<OrderItem>();
    }

    public class OrderItem
    {
        public string Name { get; set; }
    }

    public class TestContext : DbContext
    {
        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder.UseSqlServer("Server=(localdb)\\MSSQLLocalDB;Database=OwnedTypes");
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<Order>(b =>
            {
                b.OwnsMany(m => m.Items, ob =>
                {
                    ob.Property<int>("Id")
                        .IsRequired()
                        .ValueGeneratedOnAdd();

                    ob.HasForeignKey("OrderId");

                    ob.HasKey("Id", "OrderId");
                    ob.ToTable("OrderItems");

                });
            });
        }
    }
}
```

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Move relationship configuration to be chained after WithOwner
Throw when using HasOne/HasMany, Entity or Set for an owned type

Fixes #9148
Fixes #12444
Fixes #14153

### CODE REVIEW

File: src/EFCore/Metadata/Conventions/Internal/RelationshipDiscoveryConvention.cs [hunk 1]
```diff
@@ -286,19 +286,23 @@ private IReadOnlyList<RelationshipCandidate> FindRelationshipCandidates(Internal
 
                 var existingOwnership = targetEntityType?.FindOwnership();
                 if (existingOwnership != null
-                    && entityType.Model.ShouldBeOwnedType(targetClrType)
-                    && (existingOwnership.PrincipalEntityType != entityType
-                        || existingOwnership.PrincipalToDependent.Name != navigationInfo.GetSimpleMemberName()))
+                    && entityType.Model.ShouldBeOwnedType(targetClrType))
                 {
-                    return configurationSource.HasValue
+                    if (existingOwnership.PrincipalEntityType != entityType
```
Comment (smitpatel) [inline src/EFCore/Metadata/Conventions/Internal/RelationshipDiscoveryConvention.cs]: Merge ifs? anything wrong with previous?

Comment (smitpatel) [reply src/EFCore/Metadata/Conventions/Internal/RelationshipDiscoveryConvention.cs]: could pull `owned` from below before the if

---
In reply to: [254423485](https://github.com/aspnet/EntityFramework/pull/14611#discussion_r254423485) [](ancestors = 254423485)

**Comment by smitpatel** [comment]
>         /// </summary>

This file lost nullable feature

---
Refers to: src/EFCore/Properties/CoreStrings.Designer.cs:1858 in 8f89bd4. [](commit_id = 8f89bd4ee2c67705e05cc7a908759f09890b2e2b, deletion_comment = False)

File: src/EFCore.Relational/OwnershipBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+using Microsoft.EntityFrameworkCore.Metadata.Builders;
+using Microsoft.EntityFrameworkCore.Metadata.Internal;
+using Microsoft.EntityFrameworkCore.Utilities;
+
+// ReSharper disable once CheckNamespace
+namespace Microsoft.EntityFrameworkCore
+{
+    /// <summary>
+    ///     Relational database specific extension methods for <see cref="OwnedNavigationBuilder" />.
+    /// </summary>
```
Comment (smitpatel) [inline src/EFCore.Relational/OwnershipBuilder.cs]: file name

File: test/EFCore.Design.Tests/Migrations/ModelSnapshotSqlServerTest.cs [hunk 3]
```diff
@@ -895,9 +895,9 @@ public virtual void Owned_types_are_stored_in_snapshot()
                                 eo => eo.EntityWithTwoProperties, eb =>
                                 {
                                     eb.HasKey(e => e.AlternateId).HasName("PK_Custom");
-                                    eb.HasForeignKey(e => e.AlternateId).HasConstraintName("FK_Custom");
-                                    eb.HasOne(e => e.EntityWithOneProperty)
-                                        .WithOne(e => e.EntityWithTwoProperties);
+                                    eb.WithOwner(e => e.EntityWithOneProperty)
```
Comment (smitpatel) [inline test/EFCore.Design.Tests/Migrations/ModelSnapshotSqlServerTest.cs]: yay!

**Comment by AndriySvyryd** [comment]
We'll need to add nullability support to the T4 template, @roji

**Comment by roji** [comment]
@AndriySvyryd and to so many other places :)

**Comment by AndriySvyryd** [comment]
@roji Yes, I meant that we should make the change there and not in the generated file as the latter will keep getting overriden.

**Comment by roji** [comment]
Yep, I'll take care of that.

**Comment by ajcvickers** [comment]
@AndriySvyryd Can we look at this API in the design meeting? (I'm thinking that we always say we should do some API reviews earlier when the design is fresh, and this seems like one of those cases.)

**Pull request merged**